### ISSUE TITLE
`io.fits` may calculate checksum incorrectly in the presence of VLA columns

### ISSUE THREAD

**Issue by kYwzor** [astropy/astropy#14396]
### Description

In some scenarios involving VLA columns, `io.fits` can fail the round-trip verification of checksums and datasums. As far as I can tell, this happens because the checksum is being created incorrectly, not because it's being verified incorrectly.

However, the data itself seems to be saved correctly. 

### Expected behavior

Being able to write any FITS file with checksums using `io.fits` and reading it back with `io.fits` without any issues.

### How to Reproduce

Minimal code example:

```python
from astropy.io import fits
import numpy as np

col1 = fits.Column(name='a', format='1A', array=['a'])
col2 = fits.Column(name='b', format='QD', array=[np.ones(1)])
tab = fits.BinTableHDU.from_columns(name='test', columns=[col1, col2])

tab.writeto('checksum.fits', checksum=True)

read_tab = fits.open('checksum.fits', checksum=True)

```

This results in the following warnings:

> WARNING: Checksum verification failed for HDU ('TEST', 1).
>  [astropy.io.fits.hdu.base]
> WARNING: Datasum verification failed for HDU ('TEST', 1).
>  [astropy.io.fits.hdu.base]

Running the created file through NASA's FITS File Verifier we get:
```
*** Warning: Data checksum is not consistent with  the DATASUM keyword
*** Warning: HDU checksum is not in agreement with CHECKSUM.
```
So it seems the file is being created with the incorrect checksum. However, the data in the file is being saved correctly as far as I can tell.

If `col1`'s format is changed to '12A', for some reason the issue does not occur. The issue also does not occur if only one of the columns is saved.

### Versions

Windows-10-10.0.19044-SP0
Python 3.10.10 (tags/v3.10.10:aad5f6a, Feb  7 2023, 17:20:36) [MSC v.1929 64 bit (AMD64)]
astropy 5.2.1
Numpy 1.24.2
pyerfa 2.0.0.1
Scipy 1.10.0
Matplotlib 3.6.3

**Comment by cmarmo** [comment]
Hi @kYwzor , thanks for opening the issue.
Note that
```py
from astropy.io import fits

col1 = fits.Column(name='a', format='1J', array=[3])
col2 = fits.Column(name='b', format='QD', array=[[1]])
tab = fits.BinTableHDU.from_columns(name='test', columns=[col1, col2])

tab.writeto('checksum.fits', checksum=True, overwrite=True)

read_tab = fits.open('checksum.fits', checksum=True)
```
also gives no `WARNING`.
I'm checking how the `A` format is managed when VLA are among columns.
Hope to offer a PR soon.

**Comment by kYwzor** [comment]
Hello @cmarmo, do you have any news on this? Thank you

**Comment by kYwzor** [comment]
I was hoping #17209 would somehow fix this but apparently not. Maybe @mhvk has a guess what may be causing this?

**Comment by mhvk** [comment]
I fear my contribution actually changed only the speed with which checksums are calculated... I've not much experience with FITS, but your example of `A12` succeeding where `A1` fails suggests possible problems with alignment. Ping @saimn and @astrofrog.

**Comment by pllim** [comment]
At least it will error faster? 😅

**Comment by kYwzor** [comment]
@saimn just saw #17806 and it sounded very similar to this issue (also a problem with VLA checksums when VLAs are in memory). However, I've just tested the minimal reproducible example above on the latest Astropy version on GitHub and it seems this issue still occurs. Do you know if this is related in any way?

**Comment by cmarmo** [comment]
Hi @kYwzor , I might have found the issue... see #18679. 🤞

**Comment by kYwzor** [comment]
Wow, great news, I'd love to see this bug gone. It looks like a surprisingly small change, let's hope that does the trick!

**Comment by pllim** [comment]
tl;dr -- One must choose between two different proposed fixes

* https://github.com/astropy/astropy/pull/18679

OR

* https://github.com/astropy/astropy/pull/18681

**Comment by mhvk** [comment]
I'm a bit confused how both PRs can give the same result (after finding in #18681 that my thought about how checksums can be added to later were wrong). Do they give the same checksum?

**Comment by kYwzor** [comment]
@mhvk in short, the two PRs don't give the same checksum, and that is the issue here. The confusing behavior boils down to these lines:

https://github.com/astropy/astropy/blob/36d2c417a8ad2bfaa9c3700407f04fec52a7c3b8/astropy/io/fits/hdu/table.py#L881-L891

Basically, if the table/heap data is currently loaded in memory, it calls `_calculate_datasum_with_heap()`, otherwise it calls `super()._calculate_datasum()`. Both of these functions ultimately call `_compute_checksum`, but they feed it data differently. In most situations, the resulting checksum is the same for both methods, but in very specific situations the results don't match, which is evidently a bug.

When we are writing a file, the data is in memory, so it goes through `_calculate_datasum_with_heap()`. When we are reading (verifying the checksum), the data is not yet in memory, so it goes through `super()._calculate_datasum()`. This explains why it is possible for the code to create a checksum for a file, with which it later disagrees when verifying. What #18679 does is that it's forcing the reading/verification to go through `_calculate_datasum_with_heap()`, which is the exact same function that is used when writing, and thus the results match. However, my analysis is that `_calculate_datasum_with_heap()` is actually outputting the wrong result, and thus it should be fixed so that it results in the same output as `super()._calculate_datasum()`.

To put it in simple terms, currently the code calculates checksum "A" when writing and then puts "A" in the file. Later, when reading the table, it calculates checksum "B", which doesn't match "A", so it throws out a warning.  #18679 makes it so that the file calculates "A" when reading, which matches the file and thus there is no warning. On the other hand, #18681 makes it so that the file calculates "B" when writing, which is matched when reading and thus there is no warning. If we check with fitsverify, it agrees with checksum "B".

**Comment by mhvk** [comment]
@kYwzor - thanks so much for the detailed analysis!  Ideally, I'd still fix `_compute_checksum`, perhaps to somehow keep track of the remainder... Note that since this is a private function, we can change the signature, e.g., we could return a `NamedTuple` that has the checksum so far, but also all information needed to continue the calculation.
